### DISCUSSION TITLE
2024 Graphics ideas

### DISCUSSION THREAD

**Discussion by Try** [Try/OpenGothic#568] [Ideas]
1. We need to back-port culling to older(and tile-based) GPU's, that cannot run mesh-shader. (complain here: https://github.com/Try/OpenGothic/issues/563 and in many other places)
Prototype with draw-indirect (non-indexed clusters) looks promising.  On Intel:
* cull-pass takes `0.06ms` 
* draw (landscape, all opaque) = `2.34ms`
* no a-test part and no-second pass

Putting it together: 2-pass culling for landscape work well in bindless-hardware; can be scaled down to run "bindfull", with one indirect draw per material (338 draws for landscape).
- [ ] How to handle level objects? 2-passing everything might be impractical, specially on "bindfull" hardware
- [ ] But we need 2-pass for trees - those are largest occluders!
- [ ] Draw-indirect is a bit of mess in vulkan&dx - need to make sure that engine runs them in similar way
- [ ] Draw-indirect requires extra work, on engine side, to handle memory-barriers

2. Better (more complete) solution for shadows 
Point from *1 are applicable here, yet less trivial - need to do depth re-projecting.
* Considered Virtual shadows: https://ktstephano.github.io/rendering/stratusgfx/svsm yet no convinced that it's worth the memory cost.
* Irregular depth seem to be a promising for opaque: https://www.nvidia.com/en-us/geforce/news/nvidia-hfts-hybrid-frustum-traced-shadows/, yet not compatible with volumetrics.

3. Better solution for water
Tessellation is quite slow and [practically] not available on Apple. Can software rasterization (metal 3.2 afair), and tessellate water up to 1 triangle per-pixel.

**Comment by Try** [comment]
Current WIP on generalized cluster shading:
Landscape-opaque `2505` clusters
![изображение](https://github.com/Try/OpenGothic/assets/122894/dc35ca48-71dd-4dc1-b4c7-0d554780fb3f)

Level-opaque `7951` clusters
![изображение](https://github.com/Try/OpenGothic/assets/122894/501d71e9-65e9-4040-ac49-13d793430d75)

Landscape-alphacut `656` clusters
![изображение](https://github.com/Try/OpenGothic/assets/122894/42c9c42f-9f1d-486e-8c04-eef045a5d551)

Level-alphacut `4407` clusters
![изображение](https://github.com/Try/OpenGothic/assets/122894/8c4b46f1-be8c-4089-83e9-f0b9501807ed)

**Reply by Try** [reply]
This naturally requires `VK_EXT_descriptor_indexing` atm; and need a fallback on older(practically mobile) hardware

**Comment by ezamelczyk** [comment]
What about mesh shaders on Apple silicon? Right now the performance is pretty bad. Is there any possibility of improving it?

**Reply by Try** [reply]
Apparently questions is not to me here. On M1/M2 (and equivalent) they went for full software emulation. On M3 is suppose to be fixed  - yet can't be 100% sure.
Also 'clusters' here can run without mesh-shader support, as non-indexed geometry, similar to [assassins creed unity](https://advances.realtimerendering.com/s2015/aaltonenhaar_siggraph2015_combined_final_footer_220dpi.pdf) (slide 12)

**Reply by ezamelczyk** [reply]
oh that's a shame. Maybe in a few years on newer hardware it'll make sense.

**Comment by Try** [comment]
Draw indirect on DX12 :(
![изображение](https://github.com/Try/OpenGothic/assets/122894/ef75612b-32b2-46b5-a7c0-714b0765725f)

**Comment by Try** [comment]
@lmichaelis , @ezamelczyk , @Nindaleth , @thokkat 

Sorry guys for mass-tagging, yet if you have time, can you please build&test `gdr-optimizations` branch?
Basically - need to check if it doesn't crash-and-burn (similar to DX12 build) on hardware different than mine.

At this point most of game should work, except known issues:
* frame-by-frame animations on textures (water splashes)
* doesn't work without bindless (`VK_EXT_descriptor_indexing`/`Metal3-argument-buffer-tier2`/`SM5.1`)

**Reply by Nindaleth** [reply]
I've loaded up my perf measuring save in front of the Khorinis hotel. Game works, at that spot outdoor performance is >20% better and indoor performance seems unchanged. No startup crashes or visual artifacts in this extremely short test. If there's anything specific I could check (rendering of grass/torch light in caves/great outdoors where you can see very far), I definitely can, just let me know.

There is one issue I can see - when loading the game, there's the "last part of the progress bar" which takes 2 seconds on master and 32 seconds on `gdr-optimizations` branch.

**Reply by lmichaelis** [reply]
I'm not seeing any issues on my laptop's RTX 3036 either. Performance is great. I haven't noticed any major differences in loading times like @Nindaleth described. Will check on my Workstation's AMD GPU in the next couple of days and report then :)

~~Edit: RT does not work at all. It immediately crashes before loading anything!~~
Edit edit: It does work sometimes but it very flaky. Sometimes the process is killed immediately, sometimes it works. Can't rule out other system issues tho. If it starts its fine.

**Reply by Try** [reply]
> No startup crashes or visual artifacts in this extremely short test. 

This is great news. And do I remember right, you GPU is AMD? AMD is somewhat unique as it emulates `[]` operator for textures/buffers - nice to know, that it works :)

> outdoor performance is >20% better 

Hm, interesting. You got about +20%, despite mesh-shader is disabled. Maybe it's a cpu performance...

> 2 seconds on master and 32 seconds on gdr-optimizations branch.

Probably this is one of short-comings of new design - it rebuild descriptor-sets multiple times during game-load. FYI: offended place would be `DrawCommands::updateCommandUniforms`

**Reply by Try** [reply]
> Edit edit: It does work sometimes but it very flaky. Sometimes the process is killed immediately, sometimes it works

@lmichaelis,  anything in crash.log?

**Reply by lmichaelis** [reply]
> @lmichaelis, anything in crash.log?

Nope, it doesn't even get that far:

![image](https://github.com/Try/OpenGothic/assets/20187490/b08a30eb-c84d-47df-a1b0-f5f4959bec80)

**Reply by thokkat** [reply]
No issues found so far. Performance is ~10% lower on GTX 1060.

**Reply by Nindaleth** [reply]
> And do I remember right, you GPU is AMD?

Yes, still the RX 6700 XT for now.

**Reply by ezamelczyk** [reply]
On macOS I'm experiencing basically the same performance downgrade as @thokkat . Tested running from xardas' tower to the port on both versions and on average I'm getting about 10-15% less frames with gdr-optimizations. 

No idea where the bottleneck appears as the frame timings appear to be split 50/50 between cpu and gpu on both versions. Also frame pacing seems to be degraded as the master version doesn't stutter at all but on this one there are occasional hiccups.

**Reply by ezamelczyk** [reply]
I just tested with decreased resolution and without SSAO and on old version I'm running at rock solid 120 fps and on the new one it hovers around 50-70. CPU timings appear to be increased almost 100%.

**Reply by ezamelczyk** [reply]
First gameplay is on the old version and the next is on the new one - https://youtu.be/gLshMBTvASQ

In the first the fps are a little below 120 because of the built in screen recording. On the hud you can see label "Pre" means time spent on one frame on the cpu and "Gpu" means time spent on the gpu. Sum of them is frame time. You can clearly see in the second one CPU is much higher although gpu is also a little bit higher.

**Reply by ezamelczyk** [reply]
Apart from that the game seems to be running fine. No glitches or crashes.

**Reply by ezamelczyk** [reply]
I am curious why does the ingame framerate counter cap out at 80 tho.

**Reply by Try** [reply]
@ezamelczyk 
> old version I'm running at rock solid 120 fps and on the new one it hovers around 50-70. CPU timings appear to be increased almost 100%.

Thank for extra testing! small correctness, that frame time is `max(cpu_time, gpu_time)`. Looking at your video, it seem CPU time was doubled and GPU time got minor increased. 
My working theory here is to blame `useResources` on apple - this call is mandatory, yet ultra expensive - looking for workarounds here.

Based on results so far - breach is not quite ready for merge - need to fix at least apple. DX12 however can be a tech depth, don't really know how-to fix GTX performance

> I am curious why does the ingame framerate counter cap out at 80 tho.

Something else is going on in system (recording?), engine measures time between begin of frame N, and begin of frame N-1

**Reply by ezamelczyk** [reply]
> Something else is going on in system (recording?), 

Yeah that makes sense. It reports correctly without recording going on. 

Feel free to ping me any time if you need more feedback.

**Reply by Try** [reply]
> 2d2c5a4af11c5b4fee5c23b57e2175c292f92abf,  Wed Jan 31 23:01:11 2024
> refactor p3

```c++
void VisualObjects::drawShadow(Tempest::Encoder<Tempest::CommandBuffer>& cmd, uint8_t fId, int layer) {
  // return;
  auto view = SceneGlobals::VisCamera(SceneGlobals::V_Shadow0 + layer);
  for(auto i:{Material::Solid, Material::AlphaTest}) {
    drawCmd.drawCommon(cmd, fId, view, i);
    drawCmd.drawCommon(cmd, fId, view, i); // oooooooooooooooops :D
    }
  }
```

apparently game renders any shadow (and gbuffer) twice, by mistake... and still has +20% on AMD

**Reply by thokkat** [reply]
Fix brings performance on par with current rendering.

**Reply by Nindaleth** [reply]
I'm getting GPU hangs at the moment on this branch (that includes both 39a5edd1 that worked previously and also the current 682af3a7), wonder if it's caused by a Mesa update.
Master branch continues to work without any issues.

**Reply by Try** [reply]
> I'm getting GPU hangs 

This is bad... hang you mean real hang? Asking as in Vulkan there is a system-defined timeout on gpu-execution (2sec usually).
And just saying:
https://gitlab.freedesktop.org/mesa/mesa/-/issues/10360

**Reply by Nindaleth** [reply]
Maybe my terminology is imprecise and it's a GPU reset?

On the loading screen, the visual output freezes at a certain point, including the built-in FPS counter and the external MangoHUD overlay. After a few seconds my display resets (as if you disconnected and reconnected the display cable) and the same frame stays stuck on the display. Keyboard and mouse inputs don't have any effect since the initial freeze.

Out of various attempts, my system partially recovered only once from the freeze and that happened basically by the X session dropping and starting again. Here's the log from that case (no logs were produced in the other cases):

```
radv/amdgpu: The CS has been cancelled because the context is lost. This context is innocent.                                                                                                                                                
vkQueueSubmit() failed (VK_ERROR_DEVICE_LOST) object=58297232, type=3 th:15705745106772784512

---crashlog(std::terminate)---
GPU: AMD Radeon RX 6700 XT (RADV NAVI22)
#1:  - /lib64/libstdc++.so.6(+0xb4fb7) [0x7f06b4cb4fb7]
#2:  - /lib64/libstdc++.so.6(+0xb5218) [0x7f06b4cb5218]
#3:  - /home/black_fox/src/OpenGothic/build/opengothic/libTempest.so(+0x87fd3) [0x7f06b5087fd3]
#4:  - /home/black_fox/src/OpenGothic/build/opengothic/libTempest.so(+0x881fc) [0x7f06b50881fc]
#5: Tempest::Detail::VDevice::submit(Tempest::Detail::VCommandBuffer&, Tempest::Detail::VFence*) - /home/black_fox/src/OpenGothic/build/opengothic/libTempest.so(_ZN7Tempest6Detail7VDevice6submitERNS0_14VCommandBufferEPNS0_6VFenceE+0x460) [0x7f06b51035e0]
#6: MainWindow::render() - /home/black_fox/src/OpenGothic/build/opengothic/Gothic2Notr(_ZN10MainWindow6renderEv+0x354) [0x66f894]
#7: Tempest::X11Api::implProcessEvents(Tempest::SystemApi::AppCallBack&) - /home/black_fox/src/OpenGothic/build/opengothic/libTempest.so(_ZN7Tempest6X11Api17implProcessEventsERNS_9SystemApi11AppCallBackE+0x4e) [0x7f06b512170e]
#8: Tempest::X11Api::implExec(Tempest::SystemApi::AppCallBack&) - /home/black_fox/src/OpenGothic/build/opengothic/libTempest.so(_ZN7Tempest6X11Api8implExecERNS_9SystemApi11AppCallBackE+0x1b) [0x7f06b5121efb]
#9: main - /home/black_fox/src/OpenGothic/build/opengothic/Gothic2Notr(main+0x206) [0x51d4e6]
```

**Reply by lmichaelis** [reply]
A similar thing happens on my AMD rig. Whenever I start OG my window server crashes fully and does not let me log back in until I reboot. I was able to recover the following messages from syslog:

<details>
<summary>Syslog Extract</summary>

```
Feb 08 16:38:52 lmichaelis-fedora rtkit-daemon[2118]: Successfully made thread 7211 of process 7206 (/projects/@Self/_OpenGothic/build/debug/opengothic/Gothic2Notr) owned by '1000' RT at priority 20.
Feb 08 16:38:52 lmichaelis-fedora rtkit-daemon[2118]: Successfully made thread 7216 of process 7206 (/projects/@Self/_OpenGothic/build/debug/opengothic/Gothic2Notr) owned by '1000' RT at priority 1.
Feb 08 16:39:16 lmichaelis-fedora kernel: gmc_v10_0_process_interrupt: 40 callbacks suppressed
Feb 08 16:39:16 lmichaelis-fedora kernel: amdgpu 0000:2f:00.0: amdgpu: [gfxhub] page fault (src_id:0 ring:24 vmid:5 pasid:32772, for process Main thread pid 7206 thread Main thread pid 7206)
Feb 08 16:39:16 lmichaelis-fedora kernel: amdgpu 0000:2f:00.0: amdgpu:   in page starting at address 0x0000ffff3f00b000 from client 0x1b (UTCL2)
Feb 08 16:39:16 lmichaelis-fedora kernel: amdgpu 0000:2f:00.0: amdgpu: GCVM_L2_PROTECTION_FAULT_STATUS:0x00501430
Feb 08 16:39:16 lmichaelis-fedora kernel: amdgpu 0000:2f:00.0: amdgpu:          Faulty UTCL2 client ID: SQC (data) (0xa)
Feb 08 16:39:16 lmichaelis-fedora kernel: amdgpu 0000:2f:00.0: amdgpu:          MORE_FAULTS: 0x0
Feb 08 16:39:16 lmichaelis-fedora kernel: amdgpu 0000:2f:00.0: amdgpu:          WALKER_ERROR: 0x0
Feb 08 16:39:16 lmichaelis-fedora kernel: amdgpu 0000:2f:00.0: amdgpu:          PERMISSION_FAULTS: 0x3
Feb 08 16:39:16 lmichaelis-fedora kernel: amdgpu 0000:2f:00.0: amdgpu:          MAPPING_ERROR: 0x0
Feb 08 16:39:16 lmichaelis-fedora kernel: amdgpu 0000:2f:00.0: amdgpu:          RW: 0x0
Feb 08 16:39:26 lmichaelis-fedora kernel: [drm:amdgpu_job_timedout [amdgpu]] *ERROR* ring gfx_0.0.0 timeout, signaled seq=38650, emitted seq=38652
Feb 08 16:39:26 lmichaelis-fedora kernel: [drm:amdgpu_job_timedout [amdgpu]] *ERROR* Process information: process Main thread pid 7206 thread Main thread pid 7206
Feb 08 16:39:26 lmichaelis-fedora kernel: amdgpu 0000:2f:00.0: amdgpu: GPU reset begin!
Feb 08 16:39:26 lmichaelis-fedora kernel: amdgpu 0000:2f:00.0: amdgpu: MODE1 reset
Feb 08 16:39:26 lmichaelis-fedora kernel: amdgpu 0000:2f:00.0: amdgpu: GPU mode1 reset
Feb 08 16:39:26 lmichaelis-fedora kernel: amdgpu 0000:2f:00.0: amdgpu: GPU smu mode1 reset
Feb 08 16:39:27 lmichaelis-fedora kernel: amdgpu 0000:2f:00.0: amdgpu: GPU reset succeeded, trying to resume
Feb 08 16:39:27 lmichaelis-fedora kernel: [drm] PCIE GART of 512M enabled (table at 0x00000083FEB00000).
Feb 08 16:39:27 lmichaelis-fedora kernel: [drm] VRAM is lost due to GPU reset!
Feb 08 16:39:27 lmichaelis-fedora kernel: [drm] PSP is resuming...
Feb 08 16:39:27 lmichaelis-fedora kernel: [drm] reserve 0xa00000 from 0x83fd000000 for PSP TMR
Feb 08 16:39:27 lmichaelis-fedora kernel: amdgpu 0000:2f:00.0: amdgpu: SECUREDISPLAY: securedisplay ta ucode is not available
Feb 08 16:39:27 lmichaelis-fedora kernel: amdgpu 0000:2f:00.0: amdgpu: SMU is resuming...
Feb 08 16:39:27 lmichaelis-fedora kernel: amdgpu 0000:2f:00.0: amdgpu: smu driver if version = 0x00000040, smu fw if version = 0x00000041, smu fw program = 0, version = 0x003a5800 (58.88.0)
Feb 08 16:39:27 lmichaelis-fedora kernel: amdgpu 0000:2f:00.0: amdgpu: SMU driver if version not matched
Feb 08 16:39:27 lmichaelis-fedora kernel: amdgpu 0000:2f:00.0: amdgpu: use vbios provided pptable
Feb 08 16:39:27 lmichaelis-fedora kernel: amdgpu 0000:2f:00.0: amdgpu: SMU is resumed successfully!
Feb 08 16:39:27 lmichaelis-fedora kernel: [drm] DMUB hardware initialized: version=0x02020020
Feb 08 16:39:27 lmichaelis-fedora kernel: [drm] kiq ring mec 2 pipe 1 q 0
Feb 08 16:39:27 lmichaelis-fedora kernel: [drm] VCN decode and encode initialized successfully(under DPG Mode).
Feb 08 16:39:27 lmichaelis-fedora kernel: [drm] JPEG decode initialized successfully.
Feb 08 16:39:27 lmichaelis-fedora kernel: amdgpu 0000:2f:00.0: amdgpu: ring gfx_0.0.0 uses VM inv eng 0 on hub 0
Feb 08 16:39:27 lmichaelis-fedora kernel: amdgpu 0000:2f:00.0: amdgpu: ring comp_1.0.0 uses VM inv eng 1 on hub 0
Feb 08 16:39:27 lmichaelis-fedora kernel: amdgpu 0000:2f:00.0: amdgpu: ring comp_1.1.0 uses VM inv eng 4 on hub 0
Feb 08 16:39:27 lmichaelis-fedora kernel: amdgpu 0000:2f:00.0: amdgpu: ring comp_1.2.0 uses VM inv eng 5 on hub 0
Feb 08 16:39:27 lmichaelis-fedora kernel: amdgpu 0000:2f:00.0: amdgpu: ring comp_1.3.0 uses VM inv eng 6 on hub 0
Feb 08 16:39:27 lmichaelis-fedora kernel: amdgpu 0000:2f:00.0: amdgpu: ring comp_1.0.1 uses VM inv eng 7 on hub 0
Feb 08 16:39:27 lmichaelis-fedora kernel: amdgpu 0000:2f:00.0: amdgpu: ring comp_1.1.1 uses VM inv eng 8 on hub 0
Feb 08 16:39:27 lmichaelis-fedora kernel: amdgpu 0000:2f:00.0: amdgpu: ring comp_1.2.1 uses VM inv eng 9 on hub 0
Feb 08 16:39:27 lmichaelis-fedora kernel: amdgpu 0000:2f:00.0: amdgpu: ring comp_1.3.1 uses VM inv eng 10 on hub 0
Feb 08 16:39:27 lmichaelis-fedora kernel: amdgpu 0000:2f:00.0: amdgpu: ring kiq_0.2.1.0 uses VM inv eng 11 on hub 0
Feb 08 16:39:27 lmichaelis-fedora kernel: amdgpu 0000:2f:00.0: amdgpu: ring sdma0 uses VM inv eng 12 on hub 0
Feb 08 16:39:27 lmichaelis-fedora kernel: amdgpu 0000:2f:00.0: amdgpu: ring sdma1 uses VM inv eng 13 on hub 0
Feb 08 16:39:27 lmichaelis-fedora kernel: amdgpu 0000:2f:00.0: amdgpu: ring sdma2 uses VM inv eng 14 on hub 0
Feb 08 16:39:27 lmichaelis-fedora kernel: amdgpu 0000:2f:00.0: amdgpu: ring sdma3 uses VM inv eng 15 on hub 0
Feb 08 16:39:27 lmichaelis-fedora kernel: amdgpu 0000:2f:00.0: amdgpu: ring vcn_dec_0 uses VM inv eng 0 on hub 8
Feb 08 16:39:27 lmichaelis-fedora kernel: amdgpu 0000:2f:00.0: amdgpu: ring vcn_enc_0.0 uses VM inv eng 1 on hub 8
Feb 08 16:39:27 lmichaelis-fedora kernel: amdgpu 0000:2f:00.0: amdgpu: ring vcn_enc_0.1 uses VM inv eng 4 on hub 8
Feb 08 16:39:27 lmichaelis-fedora kernel: amdgpu 0000:2f:00.0: amdgpu: ring vcn_dec_1 uses VM inv eng 5 on hub 8
Feb 08 16:39:27 lmichaelis-fedora kernel: amdgpu 0000:2f:00.0: amdgpu: ring vcn_enc_1.0 uses VM inv eng 6 on hub 8
Feb 08 16:39:27 lmichaelis-fedora kernel: amdgpu 0000:2f:00.0: amdgpu: ring vcn_enc_1.1 uses VM inv eng 7 on hub 8
Feb 08 16:39:27 lmichaelis-fedora kernel: amdgpu 0000:2f:00.0: amdgpu: ring jpeg_dec uses VM inv eng 8 on hub 8
Feb 08 16:39:27 lmichaelis-fedora kernel: amdgpu 0000:2f:00.0: amdgpu: recover vram bo from shadow start
Feb 08 16:39:27 lmichaelis-fedora kernel: amdgpu 0000:2f:00.0: amdgpu: recover vram bo from shadow done
Feb 08 16:39:27 lmichaelis-fedora kernel: amdgpu 0000:2f:00.0: amdgpu: GPU reset(2) succeeded!
Feb 08 16:39:27 lmichaelis-fedora kernel: [drm] Skip scheduling IBs!
Feb 08 16:39:27 lmichaelis-fedora kernel: [drm] Skip scheduling IBs!
Feb 08 16:39:27 lmichaelis-fedora kernel: [drm] Skip scheduling IBs!
Feb 08 16:39:27 lmichaelis-fedora kernel: [drm] Skip scheduling IBs!
Feb 08 16:39:27 lmichaelis-fedora kernel: [drm] Skip scheduling IBs!
Feb 08 16:39:27 lmichaelis-fedora kernel: [drm] Skip scheduling IBs!
Feb 08 16:39:27 lmichaelis-fedora kernel: [drm] Skip scheduling IBs!
Feb 08 16:39:27 lmichaelis-fedora kernel: [drm] Skip scheduling IBs!
Feb 08 16:39:27 lmichaelis-fedora kernel: [drm] Skip scheduling IBs!
Feb 08 16:39:27 lmichaelis-fedora kernel: [drm] Skip scheduling IBs!
Feb 08 16:39:27 lmichaelis-fedora kernel: [drm] Skip scheduling IBs!
Feb 08 16:39:27 lmichaelis-fedora kernel: [drm] Skip scheduling IBs!
Feb 08 16:39:27 lmichaelis-fedora kernel: [drm] Skip scheduling IBs!
Feb 08 16:39:27 lmichaelis-fedora kernel: [drm] Skip scheduling IBs!
Feb 08 16:39:27 lmichaelis-fedora audit[4045]: ANOM_ABEND auid=1000 uid=1000 gid=1000 ses=3 subj=unconfined_u:unconfined_r:xserver_t:s0-s0:c0.c1023 pid=4045 comm="Xwayland:cs0" exe="/usr/bin/Xwayland" sig=6 res=1
```

</details>

When I try to log back in after the crash but before a reboot I get 

```
Feb 08 16:39:29 lmichaelis-fedora abrt-notification[7380]: [🡕] Process 4045 (Xwayland) crashed in amdgpu_ctx_create()
```

and then a bit later 

```
Feb 08 16:39:30 lmichaelis-fedora gnome-shell[7603]: Made thread 'KMS thread' realtime scheduled
Feb 08 16:39:30 lmichaelis-fedora gnome-shell[7603]: Device '/dev/dri/card1' prefers shadow buffer
Feb 08 16:39:30 lmichaelis-fedora gnome-shell[7603]: Added device '/dev/dri/card1' (amdgpu) using atomic mode setting.
Feb 08 16:39:30 lmichaelis-fedora gnome-shell[7603]: Created gbm renderer for '/dev/dri/card1'
Feb 08 16:39:30 lmichaelis-fedora gnome-shell[7603]: Boot VGA GPU /dev/dri/card1 selected as primary
Feb 08 16:39:30 lmichaelis-fedora gnome-shell[7603]: Obtained a high priority EGL context
Feb 08 16:39:30 lmichaelis-fedora gnome-shell[7603]: Obtained a high priority EGL context
Feb 08 16:39:30 lmichaelis-fedora abrt-notification[7790]: [🡕] Process 3381 (gnome-shell) crashed in amdgpu_ctx_create()
```

To verify I isn't a general system issue I launched BG3 which ran fine. The crash happens regardless of whether I enable or disable raytracing.

**Reply by Try** [reply]
> Maybe my terminology is imprecise and it's a GPU reset?

Important part here is `VK_ERROR_DEVICE_LOST`, that is usually called as-is "device lost". This is umbrella error code in Vulkan meaning "something went wrong". We have only one for GPU-side errors in Vulkan. DX12 has 2 - timeout is dedicated error-code.

Can be anything - infinite loop, invalid access to array, bad alignment, null, bad driver, etc. Today and yesterday been debugging similar error on IntelUHD - maybe it a shader/descriptor-set mismatch.

Plan so far is to de-risk it a little and add new command-line key `-bl 0` - to disable bindless. Hopefully by the end of week new build will be ready for testing.

> To verify I isn't a general system issue I launched BG3 which ran fine

Haven't traced BG3 :) Yet, probably it not as crazy on shading, and can work with draw-spam just fine. Here for GDR build we have 2 things that make life hard for hardware - GPU-driven visibility testing (lock-free + CS) and bindless (landscape is 2 draws pretty-much)

**Comment by Try** [comment]
New testing version is available (178dd6c2):

Animated textures and water tesselation is back. Tested on my end RTX3070/IntelUHD/MacM1 - no validation errors, no crashes.
Mesh shader is ~15% faster on PC and still slower on apple. Gap for mesh-rendering on apple however tighter than it was - they seem to have large fixed draw-command overhead, for mesh-shading.

DX12 still broken - if everything else is fine, probably it's OK to merge it as-is and solve DX problem later.
Game still has chance of stutter, if need to reallocate memory for anything - normally, if new type of material appear on a scene - TODO

Trouble shooting flags:
* `-bl 1` - command-line option, to toggle bindless on and off.
* `-ms 1` - enable or disable mesh-shader.

**Reply by lmichaelis** [reply]
Here's the results on my RX 6800 XT. I tested using a release build.

| rt | ms | bl | Comment | FPS Outdoors[^1] | FPS Indoors[^2] |
|----|----|----|---------|--------------|-------------|
| 0  | 0  | 0  | Works | ~115 - ~140 | ~140 - ~170 |
| 1  | 0  | 0  | Works | ~160 | ~100 |
| 1  | 1  | 0  | Works | ~110 | ~95 |
| 1  | 0  | 1  | Works, but has a lot of stutters when starting a new game | ~160 | ~110 |
| 1  | 1  | 1  | Works, but has a lot of stutters when starting a new game | ~160 | ~110 |

I did not notice any visual artifacts during testing. Apart from stutters when starting a new save file I didn't notice any. The FPS numbers seem very unreliable so perhaps its CPU bound now. The Gothic data files are stored on an SSD.

[^1]: Tested inside Bosper's shop
[^2]: Tested outside Bosper's shop looking towards the harbour

**Reply by Try** [reply]
what would be you opinion?
should we merge it as-is and switch to PR work or this branch requires more work?

**Reply by lmichaelis** [reply]
Looks good to merge I'd say. You should work out those stutters before the next release though so that new users don't immediately have a bad experience :)

**Reply by Nindaleth** [reply]
Current branch state is also working for me (6700 XT) when loading a save game, no crashes or stutters on Mesa git. `-rt 1 -bl 1 -ms 1`, loading takes around 10 seconds longer than on master. My usual testing location, `master` -> `gdr-optimizations` outdoors 92 -> 125 FPS, indoors 40 -> 42 FPS.

**Reply by Try** [reply]
Added quick fix for stutters: 0a4c1bfa - not a 100% solution, but might be just good enough

Do last testing before merge, GPU timeline:
![изображение](https://github.com/Try/OpenGothic/assets/122894/123b6420-bdcd-42e0-948b-47c4203e677c)

CPU flamegraph
![изображение](https://github.com/Try/OpenGothic/assets/122894/ac5f1c92-68e6-4c87-a49e-c82865407b8c)

Wind animation on trees takes ~15%, what is absolutely delightful... 
InstanceStorage - ~10% is to copy (std::memcpy) updated game state to GPU memory. 

Still more optimization work for future.

**Comment by KirillAlekseeenko** [comment]
Hello!
I want to bring up AA problem again. Since the artistic content in Gothic is not complex, probably the main source of aliasing is geometry.
That's why we could prefer MSAA over TAA (some popular solution like fsr2 or 3) , because MSAA combats geometry aliasing and doesnt have artifacts such as flickering or ghosting.
What do you think about this implementation below?
1. Depth-prepass with MSAA 4x
2. Resolve to additional depth buffer , since we need it to construct HiZ and later
3. G-buffer pass with MSAA 4x
4. Mark 'complex' pixels in stencil buffer
5. For each current g-buffer lighting pass:
       // we use stencil buffer from previous step below to avoid complex pixels or simple pixels when we dont need them
       first pass: calculate lighting for simple (1-sample) 
       second pass: calculate lighting for complex pixels

Seems that it's too expensive, but i couldn't come up with better solution yet

**Reply by Try** [reply]
What you proposing is well... normal MSAA based solution. Prepass is somewhat out of place (HiZ can be sourced from non-MSAA depth).
Would cause issues with 2-phase gbuffer . 2-phase rendering is something what most GPU-driven engines do today: 1st pass: use old HiZ, -> 1.1 refresh HiZ -> 2. Fixup pass.

I've mention my thought about AA sometime before. MSAA gbuffer is expensive memory-wise, for both store and drawing into. specially on 2k displays. 

I don't see great option to do AA today, so just do other stuff for now. Eventually we should expect Vulkan to support tile-shader, and we already got multi-sample to single-sample rendering. What we can do then:
1. Draw MSAA4 into on-chip memory
2. Run tile-shader to do "clever" lighting
3. Store as result and GBuffer as non-MSAA, to compose fog/transparency on top.

Another option is [Quincunx Antialiasing](https://blog.demofox.org/2015/04/22/quincunx-antialiasing/), - requires MSAA2 + custom sample locations + pain in fragment shader.

And lastly MSAA visibility buffer can be an option for hi-end gpu's. We don't even have to use hardware-MSAA here, just 4x resolution, and write depth(32f) + clusterId(24bit) + primitiveId(7bit). Then compute-shader magic to resolve it to color and light-up.
Cons: to do it we have to relay on bindless and fast gpu-memory, for storage-images.

**Comment by Try** [comment]
Spent quite a time tuning sky/fog against path-traced reference:

// 20:00
![изображение](https://github.com/Try/OpenGothic/assets/122894/8b486e55-79df-4dd3-81c5-f0717fcc5c53)

// 8:00
![изображение](https://github.com/Try/OpenGothic/assets/122894/8fe68a5a-67d6-4c54-8f0b-166fa60a62f9)

// 12:00
![изображение](https://github.com/Try/OpenGothic/assets/122894/89988230-1547-492e-a944-f8d961a80dbe)

This change enables to remove hack for direct-light (it was artificially dimmed) and enables better GI

**Comment by KirillAlekseeenko** [comment]
Experimented with Cmaa2 (https://www.intel.com/content/www/us/en/developer/articles/technical/conservative-morphological-anti-aliasing-20.html from MLAA family), looks less blurry than FXAA, also has a potential to integrate with MSAA (looks too complicated though, since would not be possible to resolve MSAA before CMAA2 according to their demo)

Without AA
![harborWithoutAA](https://github.com/Try/OpenGothic/assets/9151078/14417ee3-66c4-4054-a978-a548c09d3b78)
CMAA2
![harborCmaa2](https://github.com/Try/OpenGothic/assets/9151078/7533a0f5-bfa8-4fe2-8c3f-dd09aa02d560)
FXAA
![harborFxaa](https://github.com/Try/OpenGothic/assets/9151078/5e089db3-9615-41c3-b425-fe83623110c0)

Without AA
![towerWithoutAA](https://github.com/Try/OpenGothic/assets/9151078/28b0f083-7e25-43b4-af09-81e2a297cb7d)
CMAA2
![towerCmaa2](https://github.com/Try/OpenGothic/assets/9151078/18c9e15d-9840-4b83-bcdd-f9524f277cc6)
FXAA
![towerFxaa](https://github.com/Try/OpenGothic/assets/9151078/5b03e716-afc5-43b8-b1ae-d17d96353548)

**Reply by Try** [reply]
Looks quite solid! Diff is noticeable only on geometry edges.
How did you implement it? Did you put final images into Intel's demo? Asking since algorithm explanation assumes indirect-compute and we don't have it in the engine.

**Reply by Try** [reply]
Few more thing to mention:
Intel uses luminance based algorithm for edge detection. In OpenGothic we can obtain 'true' luminance for HDR linear image and potentially have better quality.
On newer NVidia gpu, there is `VK_NV_fragment_coverage_to_color` providing exact information about fragment coverage, to be used later for perfect edge detection.
In principle, if we would use visibility buffer, instead of gbuffer, then edge detection between objects would be zero cost.

**Reply by KirillAlekseeenko** [reply]
> Looks quite solid! Diff is noticeable only on geometry edges. How did you implement it? Did you put final images into Intel's demo? Asking since algorithm explanation assumes indirect-compute and we don't have it in the engine.

No, the algorithm was implemented in the engine. I'll create a pull request in Tempest with indirect dispatch. There's one more problem, I haven't figured out yet how to use the same resource as a render target and then as a UAV, so for now there's a redundant copy from the tonemapping attachment to the storage buffer cmaa2 output, since the algorithm resolves color 'in-place'.

> Intel uses luminance based algorithm for edge detection. In OpenGothic we can obtain 'true' luminance for HDR linear image and potentially have better quality.

Yeah, that would be great. For now, on screenshots, it's just a difference calculated from LDR input.

**Reply by Try** [reply]
Thanks for PR!
> I haven't figured out yet how to use the same resource as a render target and then as a UAV,

No way to do it, on purpose: UAV image is more expensive for writes, so have this bit on render-target will cause performance hit. Best approach here is to not resolve in-place, but have one image as input(RT) and separated image (UAV) as output.
